fairpy/items/saban_sethuraman_exchange_algorithm.py

```diff
@@ -0,0 +1,678 @@
+                    if len(List) > 1:
+                        flag = 1
+                        # check if List is "final" SCC
+                        for li in List:
+                            # if all node in List point doesn't point outside of the SCC.
+                            for nodefromli in graph[li]:
+                                if not List.__contains__(nodefromli):
+                                    flag = 0
+                            # if there is agent in SCC list that jealous.
+                            if owner_house.keys().__contains__(li) and not graph.has_edge(li, owner_house[li]):
+                                flag = 0
+                        # if SCC "final".
+                        if flag == 1:
+                            for i in List:
+                                graph.remove_node(i)
+                            change = 1
+                            graph = make_graph(PreferenceLists, graph)
+    return graph
+
+
+def connect_jealous_agents_to_there_best(jealous: list, graph):
+    """
+    part 1(a)
+    part 2(1.a)
+    # O(n^2)
+    this func keep only one edge between all jealous agents to the house(the house that have already edge and also
+    the house has the min "aski" value).
+    :param jealous:list with all jealous agents.
+    :param graph:diGraph with agents and houses nodes and edges that connect between them.
+    :return: update graph.
+
+
+    >>> PreferenceLists = {'1': ['a', 'b'], '2': ['b', 'a'], '3': ['d'], '4': ['c']}
+    >>> house = {'1': 'a', '2': 'b', '3': 'c', '4': 'd'}
+    >>> graph = make_graph_begin(PreferenceLists, house)
+    >>> graph = connect_jealous_agents_to_there_best([ '3', '4', 'd', 'c'], graph)
+    >>> graph = nx.to_dict_of_dicts(graph, edge_data=None)
+    >>> stringify(graph)
+    '{1:{a:{}}, 2:{b:{}}, 3:{d:{}}, 4:{c:{}}, a:{1:{}}, b:{2:{}}, c:{3:{}}, d:{4:{}}}'
+
+    >>> PreferenceLists = {'1': [{'b','c','d'}], '2': [{'a','c','d'}], '3': [{'c','d'}], '4': [{'c','d'}]}
+    >>> house = {'1': 'a', '2': 'b', '3': 'c', '4': 'd'}
+    >>> graph = make_graph_begin(PreferenceLists, house)
+    >>> graph = connect_jealous_agents_to_there_best([ '3', '4', 'd', 'c'], graph)
+    >>> graph = nx.to_dict_of_dicts(graph, edge_data=None)
+    >>> stringify(graph)
+    '{1:{b:{}, c:{}, d:{}}, 2:{a:{}, c:{}, d:{}}, 3:{c:{}}, 4:{c:{}}, a:{1:{}}, b:{2:{}}, c:{3:{}}, d:{4:{}}}'
+    """
+    # O(n)
+    # run over all the jealous agents.
+    for je in jealous:
+        MinHouse = min(graph[je].keys())
+        graphlllist = []
+        for i in graph[je]:
+            graphlllist.append(i)
+        # O(n)
+        # delete all the edges that go from jealous agents outside then MinHouse.
+        for node in graphlllist:
+            if node != MinHouse:
+                if graph.has_edge(je, node):
+                    graph.remove_edge(je, node)
+    return graph
+
+
+def connect_satisfied_agents_to_there_best(graph, owner_house: Dict, labeled: list):
```

Comment:
מה זה "there best" - האם התכוונת "their best house"?

Reply:
I change it.